nuxt3 integration docs (#1044)
intlify/vue-i18n-next: docs/guide/integrations/nuxt3.md

# Nuxt 3 integration

We will introduce Vue I18n integration for Nuxt 3.

> The following is a tutorial on set up a Nuxt 3 application from the initial configuration.

Reviewer: ```suggestion
The following is a tutorial on setting up a Nuxt 3 application from the initial configuration.
```

```ts
import { createI18n } from 'vue-i18n'

export default defineNuxtPlugin(({ vueApp }) => {
  const i18n = createI18n({
    legacy: false,
    globalInjection: true,
    locale: 'en',
    messages: {
      en: {
        hello: 'Hello, {name}!'
      }
    }
  })

  vueApp.use(i18n)
})
```

We aim to set up Vue I18n in the implementation here.

Configure up locale resources to localize a Nuxt 3 application is described in the next section.

### Run the Nuxt 3 application

Let's see if Vue I18n works with Nuxt 3.

We will edit `app.vue` of the setup Nuxt 3 application as follows:

```diff
 <template>
   <div>
-   <NuxtWelcome />
+   <h1>{{ $t('hello', { name: 'vue-i18n' }) }}</h1>
   </div>
 </template>
```

We have edited and saved, run the following command to run the Nuxt 3 application at local:

```sh
npm run dev
```

After we will run the command and access to `http://localhost:3000`, you can see the display similar to the following:

![Nuxt3 setup](/nuxt3-setup.png)

## Localize your Nuxt 3 application

So far we have been able to integrate Vue I18n into our Nuxt 3 application.
Next, we will implement language switching and import locale resources from outside.

> By implementing language switching, we can i18n our Nuxt 3 application. Also, when we will be separating locale resoruces from the source code and externalizing them, we can be localized in a separate workflow using the Localization service.

Reviewer: ```suggestion
By implementing language switching, we can i18n our Nuxt 3 application. Also, when we will be separating locale resources from the source code and externalizing them, we can be localized in a separate workflow using the Localization service.
```

In the following sections, we will work on Nuxt 3 applications that support English, French, and Japanese.

### Add language switching

We will add language switching feature to `app.vue` as follows:

```diff
 <template>
   <div>
     <h1>{{ $t('hello', { name: 'vue-i18n' }) }}</h1>
+    <form>
+      <label for="locale-select">{{ $t('language') }}: </label>
+      <select id="locale-select" v-model="$i18n.locale">
+        <option value="en">en</option>
+        <option value="fr">fr</option>
+        <option value="ja">ja</option>
+      </select>
+    </form>

> +  </div>

Reviewer: `</div>` is already in `app.vue`, not a new addition, so no plus sign is needed at the beginning of the line.

### Install `@intlify/unplugin-vue-i18n`

```sh
npm install --save-dev @intlify/unplugin-vue-i18n
```

### Configure Nuxt config

Configure `nuxt.config.ts` like the below:

```diff
 import { defineNuxtConfig } from 'nuxt'
+import { resolve, dirname } from 'node:path'
+import { fileURLToPath } from 'url'
+import VueI18nVitePlugin from '@intlify/unplugin-vue-i18n/vite'
 
 // https://v3.nuxtjs.org/api/configuration/nuxt.config
 export default defineNuxtConfig({
+  vite: {
+    resolve: {
+      alias: {
+        'vue-i18n': 'vue-i18n/dist/vue-i18n.runtime.esm-bundler.js'
+      }
+    },
+    plugins: [
+      VueI18nVitePlugin({
+        include: [
+          resolve(dirname(fileURLToPath(import.meta.url)), './locales/*.json')
+        ]
+      })
+    ]
+  }
 })
```

The bundler for Nuxt 3 is vite by default. So we will use the `vite` option here for optimization.

In `vite.resolve.alias`, set the `vue-i18n` alias to use only the Vue I18n runtime (`vue-i18n/dist/vue-i18n.runtime.esm-bundler.js`). This setting reduces the bundle size since the message compiler used by Vue I18n is not included.

> In `vite.plugins`, the plugin for `@intlify/unplugin-vue-i18n` is configured. As an option for this plugin, the `include` option specifies locale resources in json format placed in the `locales` directory. This allows `@intlify/unplugin-vue-i18n` to pre-compile locale resoruces at bundle time using Vue I18n message compiler internally. This improves the translation performance of Vue I18n and consequently the rendering performance of Nuxt 3 applications.

Reviewer 2: ```suggestion
In `vite.plugins`, the plugin for `@intlify/unplugin-vue-i18n` is configured. As an option for this plugin, the `include` option specifies locale resources in json format placed in the `locales` directory. This allows `@intlify/unplugin-vue-i18n` to pre-compile locale resources at bundle time using Vue I18n message compiler internally. This improves the translation performance of Vue I18n and consequently the rendering performance of Nuxt 3 applications.
```

### Inside of bundling with optimization

When you have finished the setup, let’s run `npm run dev` to check it out!

When you will access to `http://localhost:3000`, the behavior of the Nuxt 3 application remains the same, but there is a change in the banding of the Nuxt 3 application.

The following is a comparison of bundle sizes measured in the network tab of devtools with and without `vite.resolve.alias`:

![Reduce bundle size](/nuxt3-reduce-bundle-size.png)

The area highlighted in blue is the code bundled by vite.

`vite.resolve.alias` setting specifies a runtime-only module for Vue I18n, which reduces the bundle size.

Also, you can see the changing in the bandling of locale resources.

Code for locale resources, depending on whether or not the `@intlify/unplugin-vue-i18n` plugin to `vite.plugins` is set. Below:

![Pre-compile](/nuxt3-pre-compile.png)

> Without the `@intlify/unplugin-vue-i18n` plugin to `vite.plugins`, locale resources are banded as **json**, but with this plugin set, locale resources are **converted from json to JavaScript code** for banding.

Reviewer 2: ```suggestion
Without the `@intlify/unplugin-vue-i18n` plugin to `vite.plugins`, locale resources are banded as **json**, but with this plugin set, locale resources are **converted from json to JavaScript code** for bandwidth.
```

![Pre-compile](/nuxt3-pre-compile.png)

> Without the `@intlify/unplugin-vue-i18n` plugin to `vite.plugins`, locale resources are banded as **json**, but with this plugin set, locale resources are **converted from json to JavaScript code** for bandwidth.

Reviewer: ```suggestion
Without the `@intlify/unplugin-vue-i18n` plugin to `vite.plugins`, locale resources are bundled as **json**, but with this plugin set, locale resources are **converted from json to JavaScript code** for bandwidth.
```

> In this tutorial,  Nuxt 3 application is small, so we can not enough experience the performance of this optimization, but as the application gets larger, it will benefit from this optimization.

Reviewer 2: ```suggestion
In this tutorial,  the Nuxt 3 application is small, so we can not enough experience the performance of this optimization, but as the application gets larger, it will benefit from it.
```

### Example codes

> You can get the code for the tutorial so far [here](https://github.com/intlify/vue-i18n-next/tree/master/examples/framework/nuxt3).

Reviewer 2: ```suggestion
You can get the code for the tutorial below on [examples/framework/nuxt3](https://github.com/intlify/vue-i18n-next/tree/master/examples/framework/nuxt3).
```

Can we also move the "Example" section above "Setup vue-i18n on Nuxt 3 Application" ?

Author: Sure! 👍

 // https://v3.nuxtjs.org/api/configuration/nuxt.config
 export default defineNuxtConfig({
+  vite: {
+    resolve: {
+      alias: {
+        'vue-i18n': 'vue-i18n/dist/vue-i18n.runtime.esm-bundler.js'
+      }

> +    },

Author: @Atinux 
remove `vite.resolve.alias` no longer, so It's set in `@intlify/unplugin-vue-i18n`.
https://github.com/intlify/bundle-tools/releases/tag/unplugin-vue-i18n%400.3.0
